azurerm/data_source_batch_account.go

```diff
@@ -0,0 +1,71 @@
+package azurerm
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func dataSourceArmBatchAccount() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceArmBatchAccountRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAzureRMBatchAccountName,
+			},
+			"resource_group_name": resourceGroupNameDiffSuppressSchema(),
+			"location":            locationForDataSourceSchema(),
+			"storage_account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"pool_allocation_mode": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"tags": tagsForDataSourceSchema(),
+		},
+	}
+}
+
+func dataSourceArmBatchAccountRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).batchAccountClient
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	ctx := meta.(*ArmClient).StopContext
+	resp, err := client.Get(ctx, resourceGroup, name)
+
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			return fmt.Errorf("Error: Batch account %q (Resource Group %q) was not found", name, resourceGroup)
+		}
+		return fmt.Errorf("Error making Read request on AzureRM Batch account %q: %+v", name, err)
+	}
+
+	d.SetId(*resp.ID)
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resourceGroup)
+
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	if autoStorage := resp.AutoStorage; autoStorage != nil {
+		d.Set("storage_account_id", autoStorage.StorageAccountID)
+	}
+
+	d.Set("pool_allocation_mode", resp.PoolAllocationMode)
```

Comment:
given these properties are present within the `properties` block within the JSON response - we should retrieve these via the nested properties block; for example:

```
if props := resp.AccountProperties; props != nil {
  if autoStorage := props.AutoStorage; autoStorage != nil {
    d.Set("storage_account_id", autoStorage.StorageAccountID)
  }
  d.Set("pool_allocation_mode", props.PoolAllocationMode)
}
```